src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -5069,6 +5069,39 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         ASR::expr_t* assign_asr_target_copy = assign_asr_target;
         this->visit_expr(*x.m_targets[0]);
         assign_asr_target = ASRUtils::EXPR(tmp);
+        //Construction-While-Assigning Problem
+        AST::exprType value_type_ast = (x.m_value)->type;
+        AST::exprType objects_to_check[3] { AST::exprType::Set,  // just check these 3 for now.
+        									AST::exprType::Dict,
+        									AST::exprType::List };  
+        for(AST::exprType &exp : objects_to_check){
+            if (exp == value_type_ast){
+                ASR::ttype_t *target_type = ASRUtils::expr_type(assign_asr_target);
+                try{
+                    this->visit_expr(*x.m_value);
+                }
+                catch (SemanticError &error){
+                    ASR::expr_t *value_ASR_after_down_cast = ASRUtils::EXPR(tmp);
+                    ASR::ttype_t *value_type = ASRUtils::expr_type(value_ASR_after_down_cast);
+                    if (!ASRUtils::check_equal_type(target_type, value_type)){
+                        std::string ltype = ASRUtils::type_to_str_python(target_type);
+                        std::string rtype = ASRUtils::type_to_str_python(value_type);
+                        diag.add(diag::Diagnostic(
+                                "Type mismatch in assignment, the types must be compatible",
+                                diag::Level::Error, diag::Stage::Semantic, {
+                                        diag::Label("type mismatch ('" + ltype + "' and '" + rtype + "')",
+                                                    {assign_asr_target->base.loc, value_ASR_after_down_cast->base.loc})
+                                })
+                        );
+                        throw SemanticAbort();
+                    }
+                }
+            }
+
+        }
+
```

Comment:
I am unsure if this is a good approach. We already catch the `SemanticError()` in `body_visitor()`.

Reply:
Thanks for this PR @assem2002. I appreciate your work. The changes in this PR look good but they seem to be more like hard coded for the specific test and might not be needed at the moment.

I think a better (and more accurate) way to resolve the issue https://github.com/lcompilers/lpython/issues/2042 is to implement error recovery mechanisms (which may/might be a larger issue to be dealt with at the moment).